penrose-web/src/engine/EngineUtils.ts

```diff
@@ -306,3 +314,391 @@ export const makeTranslationDifferentiable = (trans: any): Translation => {
+              tag: "PropertyPath",
+              name: p.name, // Note use of alias
+              field: p.field,  // Note use of alias
+              property: path.property
+            }, expr, trans);
+          }
+        }
+
+        // TODO (error)
+        const err = `Err: Sub obj '${name.contents.value}' does not have GPI '${field.value}'; cannot add property '${prop.value}'`;
+        throw Error(err);
+
+      } else if (fieldRes.tag === "FGPI") {
+        const [, properties] = fieldRes.contents;
+        properties[prop.value] = expr;
+
+        // COMBAK revert
+        if (prop.value === "w") {
+          // debugger;
+        }
+
+        return trans;
+      } else { throw Error("unexpected tag"); }
+    }
+
+    case "AccessPath": {
+      const [innerPath, indices] = [path.path, path.indices];
+
+      switch (innerPath.tag) {
+        case "FieldPath": {
+          // a.x[0] = e
+          [name, field] = [innerPath.name, innerPath.field];
+          const res = trans.trMap[name.contents.value][field.value];
+          if (res.tag !== "FExpr") {
+            throw Error("did not expect GPI in vector access");
+          }
+          const res2: TagExpr<IVarAD> = res.contents;
+          // Deal with vector expressions
+          if (res2.tag === "OptEval") {
+            const res3: Expr = res2.contents;
+            if (res3.tag !== "Vector") {
+              throw Error("expected Vector");
+            }
+            const res4: Expr[] = res3.contents;
+
+            if (expr.tag === "OptEval") {
+              res4[exprToNumber(indices[0])] = expr.contents;
+            } else if (expr.tag === "Done") {
+              res4[exprToNumber(indices[0])] = floatValToExpr(expr.contents);
+            } else {
+              throw Error("unexpected pending val");
+            }
+
+            return trans;
+          } else if (res2.tag === "Done") {
+            // Deal with vector values
+            const res3: Value<IVarAD> = res2.contents;
+            if (res3.tag !== "VectorV") {
+              throw Error("expected Vector");
+            }
+            const res4: IVarAD[] = res3.contents;
+
+            if (expr.tag === "Done" && expr.contents.tag === "FloatV") {
+              res4[exprToNumber(indices[0])] = expr.contents.contents;
+            } else {
+              throw Error("unexpected val");
+            }
+
+            return trans;
+          } else {
+            throw Error("unexpected tag");
+          }
+        }
+
+        case "PropertyPath": {
+          const ip = innerPath as IPropertyPath;
+          // a.x.y[0] = e
+          [name, field, prop] = [ip.name, ip.field, ip.property];
+          const gpi = trans.trMap[name.contents.value][field.value] as IFGPI<VarAD>;
+          const [, properties] = gpi.contents;
+          const res = properties[prop.value];
+
+          if (res.tag === "OptEval") {
+            // Deal with vector expresions
+            const res2 = res.contents;
```

Comment:
Check tag first and destructure next. Probably more descriptive names + type annotations will make the code more readable.